lib/bsdlib/bsd_os.c

```diff
@@ -47,12 +47,61 @@
 static const nrfx_uarte_t uarte_inst = NRFX_UARTE_INSTANCE(1);
 #endif
 
+#define NUM_SLEEPING_THREADS 10
+
 void IPC_IRQHandler(void);
 
+static inline int thread_id_hash(u32_t thread_id)
+{
+	/* TODO: A better idea on how to hash thread IDs without collision.
+	 * A linked list or an array to fill would introduce sync issues.
+	 */
+	return thread_id % NUM_SLEEPING_THREADS;
+}
+
+typedef struct {
+	u32_t thread_id;
+	bool sleeping;
+	struct k_sem sem;
+} sleeping_thread_t;
+
+sleeping_thread_t sleeping_threads[NUM_SLEEPING_THREADS];
+
+static void sleeping_threads_init(void)
+{
+	for (int i = 0; i < NUM_SLEEPING_THREADS; i++) {
+		k_sem_init(&sleeping_threads[i].sem, 0, 1);
+		sleeping_threads[i].thread_id = 0;
+		sleeping_threads[i].sleeping = false;
+	}
+}
+
+static void all_sleeping_threads_wake(void)
+{
+	for (int i = 0; i < NUM_SLEEPING_THREADS; i++) {
+		if (sleeping_threads[i].sleeping) {
+			k_sem_give(&sleeping_threads[i].sem);
+		}
+	}
+}
+
 int32_t bsd_os_timedwait(uint32_t context, uint32_t timeout)
 {
-	/* TODO: to be implemented */
-	return 0;
+	int converted_timeout = K_MSEC(timeout);
+
+	if (timeout == 0) {
+		converted_timeout = K_FOREVER;
+	}
+
+	int i = thread_id_hash(context);
```

Comment:
How this `thread_id_hash` function is implemented looks seriously wrong to me.

Following the call stack of `bsd_os_timedwait` we can see that `context` parameter in this case is nothing else than a pointer to an array of `struct nrf_pollfd` created in `nrf91_socket_offload_poll`. So after applying this `thread_id_hash` on this address, we end up with a more or less random index `i`.

Now, this could work if we only allow to call `poll` from a single thread, but that's not acceptable, nor I don't think it was intention, seeing how it is implemented. If we call `poll` from two or more threads, it will work by chance. If the same index is calculated for multiple threads, only one of them will be woken up - others will stay permamently blocked (or until timeout occurs, but it's quite common to block 'forever').